src/lib.rs

```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
     }
 }
 
-/// Runs the ISPC kernel on the source image, sampling it down to the `target_width` and `target_height`. Returns the downsampled pixel data as a `Vec<u8>`.
+// Defines a line of weights. `coefficients` contains a weight for each pixel after `start`
+#[derive(Debug, Clone)]
+struct CachedWeight {
+    pub start: u32,
+    pub coefficients: Rc<Vec<f32>>,
+}
+
+pub(crate) fn calculate_weights(src: u32, target: u32, filter_scale: f32) -> Vec<CachedWeight> {
+    assert!(
+        src > target,
+        "Trying to use downsampler to upsample or perform an operation which will cause no changes"
+    );
+    // Every line of weights is based on the start and end of the line, and its "center" which has the biggest weight.
+    // These weight lines follow a pattern, so we can skip calculating some of them by caching all different line we get.
+    // For that purpose, we first determine the variables which define the line.
+    let mut variables = vec![WeightVariables::default(); target as usize];
+
+    unsafe {
+        ispc::downsample_ispc::calculate_weight_variables(
+            filter_scale,
+            src,
+            target,
+            variables.as_mut_ptr(),
+        );
+    }
+
+    let image_scale = src as f32 / target as f32;
+
+    let mut res = Vec::with_capacity(target as usize);
+
+    // We cache the weights in a map so that we can reuse them as we need.
+    // Half of the total number of weights seems like a good starting point to avoid unnecessary copies when resizing.
+    let mut reuse_heap = HashMap::<_, Rc<Vec<f32>>>::with_capacity(target as usize / 2);
+
+    for v in variables.iter() {
+        let coefficient_count = (v.src_end - v.src_start + 1.0) as u32;
+        // The unique values that define a collection of cached weights are how many pixels it includes and the distance from its start to its center.
+        // We use them to create a key based on which we reuse ones we've calculated previously.
+        let reuse_key = (
+            coefficient_count,
+            (v.src_center - v.src_start).to_ne_bytes(),
+        );
+
+        let reused = reuse_heap.get(&reuse_key);
```

Comment:
Nit: there should be a bunch of functions that help with efficiently retrieving or inserting a new item, without doing the lookup multiple times.